Rust/src/tree/node.rs

```diff
@@ -0,0 +1,257 @@
+        right: NodeKey,
+        bounding_box: BoundingBox<T>,
+        cut: Cut<T>) -> Self
+    {
+        Internal {
+            parent: None,
+            left: left,
+            right: right,
+            mass: 1,
+            bounding_box: bounding_box,
+            cut: cut,
+        }
+    }
+
+    /// Returns the key of the parent [`Internal`] node.
+    pub fn parent(&self) -> Option<usize> { self.parent }
+
+    /// Returns the node key of the left child.
+    pub fn left(&self) -> NodeKey { self.left }
+
+    /// Sets the left child by node key.
+    pub fn set_left(&mut self, left: NodeKey) { self.left = left }
+
+    /// Returns the node key of the right child.
+    pub fn right(&self) -> NodeKey { self.right }
+
+    /// Sets the right child by node key.
+    pub fn set_right(&mut self, right: NodeKey) { self.right = right }
+
+    /// Returns a reference to this node's bounding box.
+    pub fn bounding_box(&self) -> &BoundingBox<T> { &self.bounding_box }
+
+    /// Sets this node's bounding box to a new bounding box.
+    pub fn set_bounding_box(&mut self, bounding_box: BoundingBox<T>) {
+        self.bounding_box = bounding_box
+    }
+
+    /// Returns a reference to this node's random cut.
+    pub fn cut(&self) -> &Cut<T> { &self.cut }
+
+    /// Sets this node's random cut to a new random cut.
+    pub fn set_cut(&mut self, cut: Cut<T>) { self.cut = cut }
+
+    /// Returns the mass of this internal node.
+    pub fn mass(&self) -> u32 { self.mass }
+
+    /// Increments the mass at this internal node by one.
+    pub fn increment_mass(&mut self) { self.mass += 1 }
+
+    /// Decrements the mass at this internal node by one.
+    pub fn decrement_mass(&mut self) { self.mass -= 1 }
+}
+
+/// An enum type representing either an [`Internal`] node or a [`Leaf`] node.
+///
+/// Node stored in a random cut tree are all of type `Node`. The enum consists
+/// of two states: `Leaf`, which contains a [`Leaf`] type, and `Internal`,
+/// which contains and [`Internal`] type.
+///
+/// The methods defined for this enum type are mainly for convenience in working
+/// agnostically with either leaves or internal nodes.
+pub enum Node<T> {
+    Leaf(Leaf),
+    Internal(Internal<T>),
+}
+
+impl<T: RCFFloat> Node<T> {
+
+    /// Create a new leaf node.
+    ///
+    /// See [`Leaf::new`] for more information.
+    pub fn new_leaf(point: PointKey) -> Self {
+        Node::Leaf(Leaf::new(point))
+    }
+
+    /// Create a new internal node.
+    ///
+    /// See [`Internal::new`] for more information.
+    pub fn new_internal(
+        left: NodeKey,
+        right: NodeKey,
+        bounding_box: BoundingBox<T>,
+        cut: Cut<T>) -> Self
+    {
+        Node::Internal(Internal::new(left, right, bounding_box, cut))
+    }
+
+    /// Returns the key of the parent [`Internal`] node.
+    pub fn parent(&self) -> Option<NodeKey> {
+        match self {
+            Node::Leaf(n) => n.parent,
+            Node::Internal(n) => n.parent,
```

Comment:
Could we avoid some of this repetition by having Leaf and Internal both implement a Node trait?

Reply:
Oh man. Good idea. But that's going to have big impact on trees, scoring, and the Python wrapper. Will need some time to think this through.

Reply:
If you like this idea but want to put it in a later PR, we could create an issue for it and add it to the project.

Reply:
Create #170 to track this issue. Will defer until after the tree and other Rust code is shared and committed so that everyone involved can see the larger picture.